lib/tapioca/compilers/dsl/smart_properties.rb

```diff
@@ -84,10 +84,11 @@ def decorate(root, constant)
 
         sig { override.returns(T::Enumerable[Module]) }
         def gather_constants
-          all_classes.select do |c|
-            c < ::SmartProperties
-          end.reject do |c|
-            name_of(c).nil? || c == ::SmartProperties::Validations::Ancestor
+          all_modules.select do |c|
+            name_of(c) &&
+              c != ::SmartProperties::Validations::Ancestor &&
+              c < ::SmartProperties &&
+              (c.is_a?(Class) || c.singleton_class < ::SmartProperties::ModuleMethods)
```

Comment:
Actually, maybe even better:
```suggestion
              c < ::SmartProperties &&
              ::SmartProperties::ClassMethods === c
```
so that we don't create a singleton class for the constant if it didn't exist.